src/block_cache/block_chunkcache.c

```diff
@@ -544,15 +596,70 @@ __wt_chunkcache_remove(
     }
 }
 
+/*
+ * __wt_chunkcache_reconfig --
+ *     Re-configure the chunk cache.
+ */
+int
+__wt_chunkcache_reconfig(WT_SESSION_IMPL *session, const char **cfg)
+{
+    WT_CHUNKCACHE *chunkcache;
+    WT_CONFIG_ITEM cval;
+    WT_DECL_RET;
+    wt_thread_t content_mgmt_thread_tid;
+    char **pinned_objects;
+    unsigned int cnt;
+
+    chunkcache = &S2C(session)->chunkcache;
+    pinned_objects = NULL;
+    cnt = 0;
+
+    /* When reconfiguring, check if there are any modifications that we care about. */
+    if ((ret = __wt_config_gets(session, cfg + 1, "chunk_cache", &cval)) == WT_NOTFOUND)
+        return (0);
+
+    if (chunkcache->type == WT_CHUNKCACHE_UNCONFIGURED)
+        WT_RET_MSG(
+          session, EINVAL, "chunk cache reconfigure requested, but cache has not been configured");
+
+    /* !! Simple approach for now, probably cond_wait would be better here. !!*/
+    while (F_ISSET(chunkcache, WT_CHUNK_CACHE_CONTENT_THREAD_WORK))
+        __wt_yield();
+
+    WT_RET(__config_get_sorted_pinned_objects(session, cfg, &pinned_objects, &cnt));
+
+    __wt_writelock(session, &chunkcache->pinned_objects.array_lock);
+    chunkcache->pinned_objects.array = pinned_objects;
+    chunkcache->pinned_objects.entries = cnt;
+    __wt_writeunlock(session, &chunkcache->pinned_objects.array_lock);
+
+    WT_RET(__wt_thread_create(
+      session, &content_mgmt_thread_tid, __chunkcache_content_mgmt_thread, (void *)session));
+
```

Comment:
I wonder why this is being spun out to a different thread, rather than doing it in the caller's thread?  Do we consider it to take multiple seconds?  Maybe there's a million entries in the chunk cache (e.g. 10 terabytes split into 10 megabyte chunks).  At any rate I don't consider `WT_CONNECTION.reconfigure` to be a high performance path - and we'd need to document that this reconfiguring is queued, not finished, when `reconfigure` returns.  It seems to be more complex with a separate thread.

Whether or not we do it in the caller's thread though, we need to consider the possibility of multiple calls to reconfigure at the same time.  The current use of `WT_CHUNK_CACHE_CONTENT_THREAD_WORK` doesn't accomplish what we need, we could potentially end up with with a mix of two array list results:
- thread 1 and thread 2 independently get to line 631 with different local `pinned_object` lists.
- thread 1 does lines 631-634 (setting pinned_objects.array) and starts thread 3 to fix the chunks
- while thread 3 is fixing chunks, thread 2 does also sets pinned_objects.array, and thread 3 starts using the new `pinned_object` list.

Without spinning off threads this is easy to fix with a lock that includes the setting of the `pinned_objects.array` and the walking of the hash tables in `chunkcache_content_mgmt_thread`.

Reply:
Hopefully, I haven't misunderstood your comment, @ddanderson but we take a spinlock when entering the `conn->reconfigure` function, ensuring that the scenario described should not occur. The spinlock should prevent multiple threads from executing the reconfigure function simultaneously, therefore avoiding the mentioned scenario.

Reply:
@korteland, could you weigh in on why this is being handled in a separate thread instead of the callers? Considering WT_CONNECTION.reconfigure isn't considered to be a high-performance path (as @ddanderson mentioned).

Reply:
Sure! Keeping the potentially expensive operation off whatever thread is doing a reconfigure was one half of the motivation.

The other half was that this thread would eventually be used for other content management work, one big example was preloading (which is *definitely* too expensive to do during the reconfigure call, since it needs to do a lot of reads).

However, given the reduced scope of the project, I think it's acceptable to defer the "create a content management utility thread" work, effectively moving it to a later project. (It's not formally approved that preloading is getting dropped, but I'm 95% sure it'll happen. Unfortunately it's all still a bit up in the air.)

Reply:
@ddanderson, thank you for pointing out that scenario. However, it is taken care of [here](https://github.com/wiredtiger/wiredtiger/pull/9466/files#diff-f197249699295c61e6689541d22bf70a3de83e563bab8a3f2409837d15e2ae06R640), we ensure the thread has started running before `WT_CONN::reconfigure` returns.

Regardless, I agree that this approach does add complexity. For now, (considering Will's comment) I have decided to simplify the implementation and handle this in the caller's thread itself.

Reply:
Okay, I see, thanks.  I think that's a good decision for now.  I'd be happy to review again when you have that.

Reply:
Updated in the recent commits.